Support mapreduce over dimensions with SkipMissing

nalimilan opened the pull request:
~~The first commit fixes things like `mapreduce(cos, +, Union{Int,Missing}[1], dims=1)`. See https://github.com/JuliaLang/julia/issues/26709 and https://github.com/JuliaLang/julia/pull/27457. A deeper revamp is needed, but given that the current code doesn't make sense, that sounds like an improvement.~~ EDIT: first commit moved to  #28089.

The second commit adds an implementation of `mapreduce` over dimensions for `SkipMissing{<:AbstractArray}`, closely based on the existing `AbstractArray` methods. This allows doing things which currently don't work like `mapreduce(skipmissing([1 2; missing 4]), dims=1)` to compute column sums while skipping missing values.

Unfortunately, the simple idea I had of just wrapping the user-provided reduction operation in a function which skips missing values does not work, because `_mapreducedim!` calls `mapreduce_impl` for efficiency when possible, and it fails when a slice contains only missing values (cf. https://github.com/JuliaLang/julia/pull/27743). Also performance would probably be worse, at least for now.

I'm not really happy about the amount of duplication this requires, but that's the only solution I could find. Apart from the boilerplate dispatching methods, we could avoid duplicating `_mapreducedim!` and `reducedim_initarray0` but that would introduce several not-so-pretty `A isa SkipMissing` checks. So maybe maintaining two very similar sets of functions in parallel isn't so bad. It should be quite easy to apply changes to both, as long as one doesn't forget that two methods exist.

Replaces https://github.com/JuliaLang/julia/pull/27818.

andreasnoack commented:
https://github.com/JuliaLang/julia/pull/27845 might be touching some of the same code.

JeffBezanson commented on base/reducedim.jl:
```diff
@@ -116,7 +116,7 @@ function _reducedim_init(f, op, fv, fop, A, region)
         x = f(zero(T))
         z = op(fv(x), fv(x))
-        Tr = typeof(z) == typeof(x) && !isbitstype(T) ? T : typeof(z)
+        Tr = typeof(z) <: T ? T : typeof(z)
```
`z isa T`

JeffBezanson commented:
Separate PRs, please.

nalimilan commented:
Moved the first commit to https://github.com/JuliaLang/julia/pull/28089. I'll keep it here until it's merged since tests won't pass without it.

stillyslalom commented:
This is a much better implementation & interface than #27818. Cheers!

nalimilan committed:
Support mapreduce over dimensions with SkipMissing

Allows calling mapreduce and specialized functinos with the dims argument
on SkipMissing objects. The implementation is copied on the generic methods,
but missing values need to be handled directly inside functions for efficiency
and because mapreduce_impl returns a Some object which needs special handling.

nalimilan commented:
I've rebased and used the same approach as in #27845. Unfortunately, special code is still needed when the first slice contains missing values. Good to go?

Just to check the optimized plain `mapreduce` is still used:
@nanosoldier `runbenchmarks("union", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/28027#issuecomment-405986762) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1c0765c9299d6616573c285129d512e0aed113ec/3984d54_vs_5369849/report.md). cc @ararslan

nalimilan committed:
Fix ambiguity

StefanKarpinski commented:
Why does this need triage? Isn't it just an optimization?

nalimilan commented:
This one is not an optimization, it adds support for something which isn't currently possible.

StefanKarpinski commented:
Ok, that's a feature not a breaking change, so also doesn't need triage.

nalimilan commented:
But somebody needs to decide whether it's OK to merge before 0.7. I think it's important given that missing values are a major feature of that release.

ararslan commented:
If it passes tests and someone who knows things approves, I'd say merge whenever. ¯\\\_(ツ)\_/¯

nalimilan commented:
Bump. Anybody willing to review this?

nalimilan commented:
See mini Julep at https://github.com/JuliaLang/julia/issues/30606.

tkf commented on base/missing.jl:
```diff
@@ -280,6 +305,185 @@ mapreduce_impl(f, op, A::SkipMissing, ifirst::Integer, ilast::Integer) =
+            end
+        end
+    else
+        filled = fill!(similar(R, Bool), false)
```
If you use an external boolean array to keep the "stage" of reduction, using `_InitialValue` + `BottomRF` as now done in `foldl` might be simpler. However, this would need to use `Array{Union{T,Missing,_InitialValue}}` as the destination/state. It would be nice if there is an API to get `Array{Union{T,Missing}}` from `Array{Union{T,Missing,_InitialValue}}` by only re-writing the type tag part of the array. Is there such an API?

nalimilan replied:
Actually we could simply use `missing` here to indicate a slice for which we haven't found a non-missing value yet. This is because in the end the initialized array should contain `missing` only if the slice contains only `missing` values, in which case an error is thrown.

(Note that the current code just returns a single value, which is the smallest/largest value in the whole array. This relies on the assumption that all entries in the array can be compared with `<`, which isn't necessarily the case. So returning the smallest/largest value for each slice would make sense, in which case the strategy I describe above should work.)

Regarding your question, I don't think such a conversion method exists, but there's a related issue: https://github.com/JuliaLang/julia/issues/26681.

pdeffebach commented:
Bumping this, as someone mentioned it on slack. I would assume this is too big a change to get in by the 1.6 feature freeze?

briochemc commented:
Bump as I stumbled on this just now :) 

Was this given up on?
